### PR TITLE
feat: implement proper error handling for swaps

### LINKED ISSUE #20271: Valid swap proposal
Depends on https://github.com/status-im/status-mobile/issues/20340
https://www.figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=47-68307&t=X2bJJ4L19W3S1mbS-0

Covers the "Not enough gas" scenario and the happy path
image

image

### LINKED ISSUE #20342: Swap proposal error
Depends on https://github.com/status-im/status-mobile/issues/20340
https://www.figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=48-124457&t=X2bJJ4L19W3S1mbS-0
image

### LINKED ISSUE #21187: Something went wrong error is shown when decimal values are entered on the Swap Page
Perhaps it would be good to fix it together with the related issue https://github.com/status-im/status-mobile/issues/21188

### Description:
It seems the paraswap experiences issues when handling certain small decimal values on the Swap Page, leading to failed transactions or unexpected behavior.

### Steps:
1. Go to the Swap Page.
2. Enter a small value with multiple decimal places (e.g., 0.000006 USDC).
3. Attempt to confirm the swap.

### Actual result:
Paraswap struggles with certain small decimal values, leading to errors or incorrect behavior during the swap process.
image

### Expected result:
Potential fixes:
- Implement rounding for entered amounts on the Swap Page, similar to the Send Page functionality introduced in [PR#20915](https://github.com/status-im/status-mobile/pull/20915). This would ensure that values are rounded such that the final decimal place represents an amount between $0.01 and $0.10 USD. This would improve the display of small amounts and avoid issues with Paraswap and other third-party integrations.
OR
- Handle this error after merging PR of [issue](https://github.com/status-im/status-desktop/issues/16267)

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

### Logs:
[Status-debug-logs (1).zip](https://github.com/user-attachments/files/16905385/Status-debug-logs.1.zip)

### PR DESCRIPTION
**Pull request by briansztamfater**
fixes #20342 
fixes #20271
fixes #21187

### Summary

This PR implements error handling for swap proposal errors.

#### Platforms

- Android
- iOS

##### Functional

- wallet / transactions

### Steps to test

- Open Status
- Login
- Go to wallet
- Select an account
- Select Swap option
- Select a token on Select asset to pay screen
- Enter a different amounts (low amount, max amount for ETH, an amount that exceeds the max amount)
- Wait for swap proposal or error to appear
- Verify the errors are shown correctly

status: ready

### CODE REVIEW

**Comment by VolodLytvynenko** [comment]
Hi @briansztamfater, should this PR be tested now, or is it better to wait until the status changes from 'draft' to 'open'?

**Comment by briansztamfater** [comment]
Hey @VolodLytvynenko, this is almost ready, but I think it is better to first verify #21179 is correctly tested, because status-go latest changes introduced some bugs in the mobile client and would make difficult to test this PR. Once we merge #21179  I will rebase this one, solve the conflicts and send it to review.

**Comment by VolodLytvynenko** [comment]
hi @briansztamfater thank you for PR. take a look at found issues

## ISSUE 1: The "Approve" button is shown as enabled even when there isn't enough ETH to cover the fee for a non-approved ERC-20 token

### Steps:
1. Recover a user without enough ETH to cover the swap transaction fee.
2. Go to the swap page using a non-approved ERC-20 token.

### Actual result:
The "Approve" button is shown as enabled and allows navigating to the approval page, despite the insufficient ETH balance to cover the transaction fee.

https://github.com/user-attachments/assets/f54e2844-79de-4de6-893e-615e5a38c50b

### Expected result:
The "Approve" button should be disabled if there isn't enough ETH to cover the transaction fee, similar to how it works on the desktop version, or the button should not be shown at all.
image

File: src/status_im/constants.cljs
Comment (clauxx) [inline src/status_im/constants.cljs:618]: More of a code style comment, but can we have a map for the errors to make it easier on the eyes :sweat_smile:? I see that we do it is as above in mostly all cases for constants, but to me it would be a bit cleaner this way:

```clj
(def router-errors
  {:error/generic              "0"
   :error/paraswap-custom      "WPP-030"
   :error/price-timeout        "WP-037"
   :error/not-enough-liquidity "WPP-038"})

;; when asserting the error code
(= error-response-code
   (:error/not-enough-liquidity constants/router-errors))
``` 

it could also be just me, so I wonder if others have any preferences, cause going through the constants namespace gets more and more annoying recently and I think it's due to the structure of things.

cc: @ilmotta @ulisesmac is there some clojure-friendly way of doing this?

**Review by clauxx — APPROVED** [review]
Great job! :100: 

P.S. While trying to fix another bug I started doing the same thing for send/bridge, where we don't react to the errors coming from the "suggested-routes" signal other than showing a toast which leads to weird states, but parked the changes since the bug was de-scoped. I guess it would be good to add similar error-handling to send/bridge eventually cc: @vkjr @shivekkhurana

File: src/status_im/constants.cljs
Comment (briansztamfater) [reply src/status_im/constants.cljs:618]: Yeah, I started using a map and then changed to something similar as the desktop implementation. Open to suggestions on how to improve the approach :)

**Comment by briansztamfater** [comment]
@VolodLytvynenko Issue 1 should be fixed!

**Comment by VolodLytvynenko** [comment]
### PR_ISSUE 2:"com.facebook.react.bridgeJSApplicationillegalArgumentException" error is shown after entering suffient decimals several times

### Steps:
1. Go to swap
2. Enter value with sufficient decimals (in my case  WBTC 0.0000010101)

### Actual result:
"com.facebook.react.bridgeJSApplicationillegalArgumentException" error is shown

https://github.com/user-attachments/assets/60497a17-91e0-41c0-8d83-78ce313b81d7


### Expected result:
The error should be handled by proper input validation, and the application should prevent the user from entering invalid decimal values.


### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

### Logs
[Status-debug-logs.zip](https://github.com/user-attachments/files/17076686/Status-debug-logs.zip)

**Comment by VolodLytvynenko** [comment]
@briansztamfater, Question 1: should we always get the 'Paraswap error: Amount <amount> is too small' in all cases when entering a small amount, or just in particular ones?

For example, when I enter 0.0000001 USDC, I get the 'Paraswap error: Amount <amount> is too small,' but when I add one more decimal, I get 'Something went wrong. Modify swap parameters or try again later.'

Is it expected or validation 'Paraswap error: Amount <amount> is too small' should be shown in both cases?

https://github.com/user-attachments/assets/7643f3de-6d56-47c1-a876-2fe1a7728ae9

**Comment by VolodLytvynenko** [comment]
Question 2: Do we have a specific pattern for when the 'amount too small' validation should appear? For example, when I enter 6 decimals like 0.0000001 for tokens like WBTC or USDC, the validation shows up, but with DAI, I can enter many decimals like 0.000000000000000001 and don’t get any validation. Is there a reason for this inconsistency?

Case when no validation is shown:
image

Case when validation is shown:
image

**Comment by briansztamfater** [comment]
@VolodLytvynenko About the `Amount too small` error questions, it really depends on Paraswap, as that error comes directly from the provider and we cannot predict or say when an amount is too small for the provider. So I would just show the error as is and let the user modify the amount.

I also fixed the decimals error, a local validation was failing and, in reality, the user should not be able to input more decimals than the token accepts. In the case of WBTC, it should accept up to 8 decimals. For DAI, it is 18 decimals. For USDC, it is 6 decimals. That should also fix #21187, so also edited the description to link that issue as well. Let me know if this fixes issue 2.

**Comment by VolodLytvynenko** [comment]
Hi @briansztamfater Thank you for the fixes. However, [Issue 2](https://github.com/status-im/status-mobile/pull/21261#issuecomment-2363886668) still persists. It seems we can get the error even when entering 6 decimals for WBTC, like 0.0000001. I noticed this happens particularly when the account doesn’t have enough funds to cover the gas fee

### Steps:
1. Recover user with available WBTC and without ETH assets on account
2. Go to swap page
3. Enter WBTC using 6 decimals

### Actual result:

https://github.com/user-attachments/assets/c9718708-f2a4-41cb-80c9-b50df44a9be9

### Logs:
[Status-debug-logs.zip](https://github.com/user-attachments/files/17094954/Status-debug-logs.zip)

**Review by mohsen-ghafouri — APPROVED** [review]

**Comment by briansztamfater** [comment]
@VolodLytvynenko thanks for testing, it seems like an Android-only bug, I was able to reproduce it and just pushed a potential fix. I tested it locally and it seems to works fine now. Let me know how testing goes!

**Comment by VolodLytvynenko** [comment]
hi @briansztamfater thank you for PR. No issues from my side. PR is ready to be merged

**Pull request merged**
